plugin/pkg/admission/initialresources/data_source.go

```diff
@@ -0,0 +1,205 @@
+	GetUsagePercentile(kind api.ResourceName, perc int64, image string, exactMatch bool, start, end time.Time) (usage int64, samples int64, err error)
+}
+
+func newDataSource(kind string) (dataSource, error) {
+	if kind == "influxdb" {
+		return newInfluxdbSource(*influxdbHost, *user, *password, *db)
+	}
+	if kind == "gcm" {
+		return newGcmSource()
+	}
+	return nil, fmt.Errorf("Unknown data source %v", kind)
+}
+
+// TODO(piosz): rewrite this once we will migrate into InfluxDB v0.9.
+type influxdbSource struct {
+	conf *influxdb.ClientConfig
+	conn *influxdb.Client
+	// protects <conn>
+	mutex sync.RWMutex
+}
+
+func newInfluxdbSource(host, user, password, db string) (dataSource, error) {
+	conf := &influxdb.ClientConfig{
+		Host:     host,
+		Username: user,
+		Password: password,
+		Database: db,
+	}
+	source := &influxdbSource{
+		conf:  conf,
+		mutex: sync.RWMutex{},
+	}
+	go source.ensureDbConnection()
+	return source, nil
+}
+
+func ensureSeriesExists(conn *influxdb.Client, existingQueries *influxdb.Series, seriesName, contQuery string) error {
+	queryExists := false
+	for _, p := range existingQueries.GetPoints() {
+		id := p[1].(float64)
+		query := p[2].(string)
+		if strings.Contains(query, "into "+seriesName) {
+			if query != contQuery {
+				if _, err := conn.Query(fmt.Sprintf("drop continuous query %v", id), influxdb.Second); err != nil {
+					return err
+				}
+			} else {
+				queryExists = true
+			}
+		}
+	}
+	if !queryExists {
+		if _, err := conn.Query("drop series "+seriesName, influxdb.Second); err != nil {
+			return err
+		}
+		if _, err := conn.Query(contQuery, influxdb.Second); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ensureAutoscalingSeriesExist(conn *influxdb.Client) error {
+	series, err := conn.Query("list continuous queries", influxdb.Second)
+	if err != nil {
+		return err
+	}
+	if err := ensureSeriesExists(conn, series[0], cpuSeriesName, cpuContinuousQuery); err != nil {
+		return err
+	}
+	if err := ensureSeriesExists(conn, series[0], memSeriesName, memContinuousQuery); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO add connection check here and make this running all the time
+func (s *influxdbSource) ensureDbConnection() {
+	for s.conn == nil {
+		time.Sleep(30 * time.Second)
+		client, err := influxdb.NewClient(s.conf)
+		if err != nil {
+			glog.Errorf("Error while trying to create InfluxDB client: %v", err)
+			continue
+		}
+		if err := ensureAutoscalingSeriesExist(client); err != nil {
+			glog.Errorf("Error while trying to create create autoscaling series: %v", err)
+			continue
+		}
+		s.mutex.Lock()
+		s.conn = client
+		s.mutex.Unlock()
+	}
+}
+
+func (s *influxdbSource) query(query string, precision ...influxdb.TimePrecision) ([]*influxdb.Series, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	if s.conn == nil {
+		return nil, fmt.Errorf("No active connection to InfluxDB")
+	}
+	return s.conn.Query(query, precision...)
```

Comment:
Only one client can query at a time? So this is essentially a 1-item connection pool. Right?